### PR TITLE
fix(desktop): require a real witness before calling waste witnessed

### PR DESCRIPTION
**Pull request by ankit-yc**
## PR Checklist

- [x] The PR title follows our guidelines: https://github.com/YosemiteCrew/Yosemite-Crew/blob/main/CONTRIBUTING.md#commit-message-convention.
- [ ] There is an issue for the bug/feature this PR is for.
- [x] All existing tests and lints pass.

Found by an adversarial DEA compliance audit of `apps/desktop`. This is blocker 3, verified by hand and reproducible with no failure of any kind. Blockers 1, 2 and 4 are in #2520; the two PRs are independent.

## What is the current behavior?

`yc:cs-record` accepts `action: 'waste'` and validates only `drugName`, `drugClass`, `lotNumber`, `unit`, `veterinarianId`, `veterinarianName` and `quantity`. There is **no** `witnessId` or `witnessName` requirement.

The read path in `dual-witness.ts` then hardcodes the result:

```ts
witnessId: tx.witnessId || '',
witnessName: tx.witnessName || '',
witnessPinVerified: true,      // not derived from anything
```

So a controlled-substance destruction recorded with no witness at all reads back as witness-verified, and `showPmpStatus` counts it under "Witnessed waste events". That is an affirmative false compliance statement about a DEA control, produced on the normal path. No crash, no disk error, no edge case is required.

`recordWaste` and `verifyWitnessPin` are exported from `dual-witness.ts` and never called anywhere outside that file and its own tests, so the dual-witness check was dead code.

## What is the new behavior?

**A witness is required at the IPC boundary.** Actions in the new `WITNESS_REQUIRED_ACTIONS` (currently `'waste'`) must carry a non-empty `witnessId` and `witnessName`, and a veterinarian may not list themselves as their own witness.

**`witnessPinVerified` reflects a real verification.** The main process derives it from an actual `verifyWitnessPin` call, persists it on the transaction, and `txToWasteEvent` reports the persisted value. Nothing synthesises `true`.

**The renderer cannot assert verification.** `witnessPinVerified` is never read from the IPC payload; whatever the caller sends is ignored. There is a test for exactly this.

**The flag is tamper-evident.** `witnessPinVerified` is signed into the audit entry's `details`. The logbook file is not HMAC-protected, so without this someone with access to the user-data directory could flip it from `false` to `true` on disk and the read paths would accept the event as verified while `verifyAll()` and `verifyChain()` still passed. The signed entry now records what was actually verified, so the two disagree and the alteration shows.

**The dialog stops mislabelling.** `showPmpStatus` reports "Witness-verified waste events: N" and, when any exist, "Waste events WITHOUT a verified witness: N" on its own line.

### What this PR deliberately does NOT do

**It does not add a witness-enrolment channel.** An earlier revision added `yc:cs-set-witness-pin`. Review correctly identified that as worse than the gap it filled: any renderer on an allowed origin could enrol an arbitrary witness id with a PIN of its own choosing and immediately submit that PIN to `yc:cs-record`, producing `witnessPinVerified: true` with one person in the room. That manufactures precisely the false verification this PR exists to remove. It has been removed, with a test asserting the channel is not reachable. Enrolment must be bound to an authenticated second identity, which is a design question for the auth story.

`verifyWitnessPin` stays wired into the record path, so the mechanism is tested and enrolment becomes a one-line change once there is an authoritative identity to bind it to. Until then it yields `false` in practice, and the dialog says so.

**That is the fix, not a shortfall.** A truthful count of unverified destructions is what removes the false compliance statement. A verified-looking count produced by a control anyone can forge would not.

**It does not refuse unverified waste.** With no enrolment path, PIN verification cannot succeed, so rejecting would make it impossible to record any destruction at all. An unrecorded destruction is a worse DEA outcome than one recorded and truthfully marked unverified: refusing the record does not undo the disposal. This differs from `recordWaste`, which does reject before touching inventory. That inconsistency is real and should be settled in favour of rejecting once enrolment exists.

## Impact area

`apps/desktop` only: the `yc:cs-record` IPC handler, `dual-witness.ts`, the `CsTransaction` shape, and the PMP dialog. No backend, frontend or mobile changes.

Callers of `yc:cs-record` with `action: 'waste'` must now supply `witnessId` and `witnessName`. This is a deliberate breaking change to that payload: the previous contract permitted a non-compliant record.

## Known limitation, worth a follow-up

`setWitnessPin` stores witness accounts in an in-memory `Map`, so nothing survives a restart. This is pre-existing and now moot for the IPC path since enrolment is not exposed, but it has to be designed together with the authenticated enrolment flow.

## Validation performed

- `npx jest`: 967 tests, 67 suites, all pass.
- Coverage: 98.03% statements, 90.0% branches, 97.53% functions. Every touched file at or above its `dev` baseline, measured by stashing and re-running: `dual-witness.ts` 95.7% to 100% branches, `ipc-handlers.ts` 79.8% to 80.8%, `status-dialogs.ts` 88.0% to 88.5%, `controlled-substance.ts` unchanged at 90.2%, `preload.ts` 100%.
- `npx tsc --noemit -p tsconfig.test.json` and `npx eslint src tests --max-warnings=0` clean.
- Pre-push monorepo lint + type-check passed.
- **Mutation checked.** Restoring `witnessPinVerified: true` on the read path fails 3 tests; removing the witness requirement fails 3; trusting the renderer's flag fails 2; counting every waste event as witnessed in the dialog fails 1; unsigning the flag fails 1; re-exposing the enrolment channel fails 1. Restoring the fixes returns everything to green. No test passes both with and without its fix.

## Related Issue(s)

Fixes #

### CODE REVIEW

**Comment by ankit-yc** [comment]
The review found a real problem with this PR and I have removed the offending piece in 180b2eec.

**Authenticate witness enrollment before accepting a PIN (P1) - agreed, channel removed**

You are right and this was the most important comment on either PR. `yc:cs-set-witness-pin` was worse than the gap it filled: any renderer on an allowed origin could enrol an arbitrary witness id with a PIN of its own choosing and immediately submit that PIN to `yc:cs-record`, producing `witnessPinVerified: true` with one person in the room. That manufactures precisely the false verification this PR exists to remove. The channel is gone from `IPC_CHANNELS`, from the handler and from preload, and there is a test asserting it is not reachable.

`verifyWitnessPin` stays wired into the record path so the mechanism is tested and enrolment becomes a one-line change once there is an authoritative identity to bind it to.

**Bind the verification flag into the signed audit entry (P1) - agreed, fixed**

Correct, and cheap. `witnessPinVerified` is now signed into the audit entry's details. The logbook file is not HMAC-protected, so previously a flag flipped from `false` to `true` on disk would have read back as witness-verified with `verifyAll()` and `verifyChain()` still passing. The signed entry now records what was actually verified, so the two disagree and the alteration shows.

**Reject waste when PIN verification fails (P1) - not taking, with reasons**

With no enrolment path, PIN verification can never succeed, so rejecting would make it impossible to record any destruction at all. An unrecorded destruction is a worse DEA outcome than a destruction recorded and truthfully marked unverified: refusing the record does not undo the disposal, it just removes the evidence that it happened.

You correctly note this differs from `recordWaste`, which does reject before touching inventory. That inconsistency is real and I have called it out in the commit message. My view is that it should be settled in favour of rejecting - but only once enrolment exists, because until then rejecting is equivalent to disabling waste recording.

What this PR does deliver is the removal of the false statement, which is the actual blocker: an unwitnessed destruction no longer reads back as witness-verified, and the PMP dialog now reports "Waste events WITHOUT a verified witness: N" instead of counting them all as witnessed. A truthful count of unverified destructions is the fix. A verified-looking count produced by a control anyone can forge is not.

**Persist witness credentials across restarts (P2) - acknowledged, out of scope**

Correct, and it is pre-existing: `setWitnessPin` has always stored only in a process-local map. It is now moot for the IPC path since enrolment is not exposed, and it should be designed together with the authenticated enrolment flow rather than bolted on. I called this out as a known limitation in the PR description before the review and it still stands.

**Aikido: witnessId in logs**

Resolved by removing the channel - `cs_witness_enrolled` was the only place a witness id was logged. `cs_recorded` logs only the transaction id, drug name and the boolean.

**Sonar S3776**

The `cs-record` handler's validation moved into a module-level `validateCsRecordPayload`, bringing cognitive complexity back under the limit.

All changes have regression tests, each mutation checked by reintroducing the defect and confirming the specific test fails.

**Review by aupyay — DISMISSED** [review]

**Review by aupyay — DISMISSED** [review]

File: apps/desktop/src/compliance/controlled-substance.ts [hunk 1]
```diff
@@ -21,10 +21,18 @@ export interface CsTransaction {
   veterinarianName: string;
   witnessId?: string;
   witnessName?: string;
+  // Whether the witness actually proved their identity at record time. Persisted
+  // rather than inferred: a waste event read back from disk carries what was
+  // verified when it happened, never an assumption that it must have been.
+  witnessPinVerified?: boolean;
```
[the remark this answers is not in the document]
Comment (ankit-yc) [reply apps/desktop/src/compliance/controlled-substance.ts:28]: Fixed. `witnessPinVerified` is signed into the audit entry's `details` in `controlled-substance.ts` `record()`, so a flag flipped in the logbook file no longer agrees with the signed side.

Worth noting this fix alone was not sufficient, and your later comments are why: signing the flag made forgery *detectable* but nothing was checking it, and even after the read path started checking, `yc:audit-append` let a renderer obtain a signature for an entry of its own choosing. Both are now closed. Test: `waste-witness.test.ts` "the verification flag is signed into the audit entry, so flipping it on disk is detectable".

File: apps/desktop/src/core/ipc-handlers.ts [hunk 2]
```diff
@@ -147,6 +153,56 @@ export interface IpcServices {
...
+
+  // Destruction may not be recorded on one person's say-so. Without this the
+  // channel accepted action:'waste' with no witness at all, and the record still
+  // read back as witness-verified.
+  const needsWitness = WITNESS_REQUIRED_ACTIONS.includes(action as CsAction);
+  if (needsWitness) {
+    const missingWitness = requireNonEmptyStrings(d, ['witnessId', 'witnessName']);
+    if (missingWitness) return { ok: false, error: missingWitness };
+    if (d.witnessId === d.veterinarianId) return { ok: false, error: 'witness-must-differ' };
+  }
```
[the remark this answers is not in the document]
Comment (ankit-yc) [reply apps/desktop/src/core/ipc-handlers.ts:229]: Fixed, though not the way the suggestion proposed. Ids are compared case- and whitespace-insensitively so `"vet-1"` and `"vet-1 "` cannot be two people, but storage only trims: these ids come from an external identity system that may treat case as significant, and folding case on the way to disk would corrupt the identifier used to match a real person.

That asymmetry is deliberate and is documented at the call site along with what it costs. Since the comparison is only ever used to REJECT, being generous about what counts as the same person fails safe. Both halves are pinned by tests: lowercasing on persist fails the storage test, and a case-sensitive comparison fails the self-witness test.

File: apps/desktop/src/core/ipc-handlers.ts [hunk 3]
```diff
@@ -374,40 +452,41 @@ export const registerIpc = (services: IpcServices, ipc: IpcMainType = ipcMain):
...
     // confirmation for a dispense that was never written down.
     let tx: CsTransaction;
     try {
-      tx = services.controlledSubstanceLog.record(
-        data as Parameters<ControlledSubstanceLogbook['record']>[0]
-      );
+      tx = services.controlledSubstanceLog.record({
+        ...(rest as Parameters<ControlledSubstanceLogbook['record']>[0]),
+        witnessPinVerified,
+      });
```
[the remark this answers is not in the document]
Comment (ankit-yc) [reply apps/desktop/src/core/ipc-handlers.ts:491]: Fixed. The daily and range CSVs now carry an explicit `Witness Verified` column (`yes`/`no`) alongside the claimed name.

You were right that this is the same defect as the one the PR set out to remove, one file over: labelling a claimed name "Witness" with no status made an unwitnessed destruction indistinguishable from a witnessed one in the DEA export, which is exactly the false statement the PMP dialog was making. I kept the claimed name rather than blanking it, since a DEA record should preserve what was entered; the status column is what removes the false implication. Test: `cs-export.test.ts` "the CSV states whether a witness was verified".

File: apps/desktop/src/core/ipc-handlers.ts [hunk 4]
```diff
@@ -374,40 +452,41 @@ export const registerIpc = (services: IpcServices, ipc: IpcMainType = ipcMain):
...
+    if (!validation.ok) return { ok: false, error: validation.error };
+
+    // The verification flag is derived here from an actual PIN check and never
+    // taken from the payload: a renderer must not be able to assert that a
+    // witness was verified.
+    const witnessPinVerified =
+      validation.needsWitness &&
+      services.dualWitnessLog !== null &&
+      typeof d.witnessPin === 'string' &&
+      services.dualWitnessLog.verifyWitnessPin(d.witnessId as string, d.witnessPin);
```
[the remark this answers is not in the document]
Comment (ankit-yc) [reply apps/desktop/src/core/ipc-handlers.ts:469]: Fixed. `getWitnessAccount()` returns the enrolled account, and the handler replaces the caller-supplied `witnessName` with the canonical one once the PIN verifies.

Your framing is the useful part and I have kept it in the code comment: a PIN proves an ACCOUNT, not a name. Without this, a caller could pair a genuine id and PIN with any name it liked and that arbitrary value is what the compliance CSV would present. Test: `ipc-handlers.test.ts` "a verified witness is recorded under the enrolled account name".

**Review by aupyay — APPROVED** [review]
